xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,288 @@
+            helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+          }
+        }
+
+        @Override
+        public boolean canHandleEmptyAddressListFromNameResolution() {
+          return true;
+        }
+
+        @Override
+        public void shutdown() {
+          if (clusterWatcher != null) {
+            if (clusterWatcher.edsBalancer != null) {
+              clusterWatcher.edsBalancer.shutdown();
+            }
+            xdsClient.cancelClusterDataWatch(cdsConfig.name, clusterWatcher);
+          }
+        }
+
+        @Override
+        public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+          if (clusterWatcher == null) {
+            clusterWatcher = new ClusterWatcherImpl(helper, resolvedAddresses);
+            xdsClient.watchClusterData(cdsConfig.name, clusterWatcher);
+            if (oldCdsConfig != null) {
+              xdsClient.cancelClusterDataWatch(oldCdsConfig.name, oldClusterWatcher);
+            }
+            CdsLoadBalancer.this.clusterWatcher = clusterWatcher;
+          }
+          // Else handleResolvedAddresses() already called, so no-op here because the config is
+          // fixed in this balancer.
+        }
+      };
+    }
+  }
+
+  private final class ClusterWatcherImpl implements ClusterWatcher {
+
+    final Helper helper;
+    final ResolvedAddresses resolvedAddresses;
+
+    // EDS balancer for the cluster.
+    // Becomes non-null once handleResolvedAddresses() successfully.
+    // Assigned at most once.
+    @Nullable
+    LoadBalancer edsBalancer;
+
+    // LoadStatsStore for the cluster.
+    // Becomes non-null once handleResolvedAddresses() successfully.
+    // Assigned at most once.
+    @Nullable
+    LoadStatsStore loadStatsStore;
+
+    ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
+      this.helper = helper;
+      this.resolvedAddresses = resolvedAddresses;
+    }
+
+    @Override
+    public void onClusterChanged(ClusterUpdate newUpdate) {
+      channelLogger.log(ChannelLogLevel.DEBUG, "Received a CDS update: '%s'",  newUpdate);
+      checkArgument(
+          newUpdate.getLbPolicy().equals("round_robin"),
+          "The load balancing policy in ClusterUpdate '%s' is not supported", newUpdate);
+
+      final XdsConfig edsConfig = new XdsConfig(
+          /* balancerName = */ null,
+          new LbConfig(newUpdate.getLbPolicy(), ImmutableMap.<String, Object>of()),
+          /* fallbackPolicy = */ null,
+          /* edsServiceName = */ newUpdate.getEdsServiceName(),
+          /* lrsServerName = */ newUpdate.getLrsServerName());
+
+      if (loadStatsStore == null) {
+        loadStatsStore = new LoadStatsStoreImpl();
+      }
+      if (edsBalancer == null) {
+        edsBalancer = lbRegistry.getProvider(XDS_POLICY_NAME).newLoadBalancer(helper);
+      }
+      edsBalancer.handleResolvedAddresses(
+          resolvedAddresses.toBuilder()
```

Comment:
Why would you need to take "partial" of `ResolvedAddresses` for CDS policy and pass it down to EDS policy? You could just assemble a new one, put in the reference of `XdsClient` (the object pool). It's more elegant.

Reply:
In general we pass all the addresses and all the unknown attributes all the way down to child balancers, not assemble a new one with only known attributes. This applies for CDS to EDS.